amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpSourceStage.scala

```diff
@@ -12,6 +14,21 @@ import com.rabbitmq.client._
 import scala.collection.mutable
 
 final case class IncomingMessage(bytes: ByteString, envelope: Envelope, properties: BasicProperties)
+final case class CommittableIncomingMessage(
+    message: IncomingMessage,
+    callback: AsyncCallback[Unit] = new AsyncCallback[Unit] {
+      override def invoke(t: Unit): Unit = Unit
+    }
```

Comment:
The message shouldn't need to know about `AsyncCallback` just a regular function should do. After thinking about the things noted below, I think it should be callback: `(ackOrNack+data) => Future[Unit]` with the whole implementation of that function in the stage using `AsyncCallback`

Reply:
I see. So you are suggesting that I wrapped the Async callback in a plain function and pass that to the Committable message, right?

There is some logic in the Stage so it won't be completed until all the messages have been acked. If something really goes wrong I guess the messages won't be acked. I think it's the same in the akka-streams-kafka, we offer out of the box at-least-once delivery but not exactly-once delivery.

The ack method from rabbitMQ is blocking AFAIK. I tried wrapping it in a Future and using mapAsync but got the issue that I mentioned in a comment above. Maybe you have any suggestion?